translate/pb/pb_util.go

```diff
@@ -0,0 +1,163 @@
+package pb
+
+import (
+	"fmt"
+
+	"github.com/arr-ai/arrai/rel"
+	"github.com/arr-ai/arrai/tools"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// decodeFileDescriptor parses protobuf definition file and decodes to file descriptor
+func decodeFileDescriptor(definition []byte) (protoreflect.FileDescriptor, error) {
+	fdSet := new(descriptorpb.FileDescriptorSet)
+	if err := proto.Unmarshal(definition, fdSet); err != nil {
+		return nil, err
+	}
+
+	// We know protoc was invoked with a single .proto file only
+	file1 := fdSet.GetFile()[0]
+
+	// Initialize the File descriptor object
+	fd1, err := protodesc.NewFile(file1, protoregistry.GlobalFiles)
+	if err != nil {
+		return nil, err
+	}
+
+	return fd1, nil
+}
+
+// WalkThroughMessageToBuildValue walks through protobuf message and build a value whose type is rel.Value
+func WalkThroughMessageToBuildValue(val protoreflect.Value) (rel.Value, error) {
+	switch message := val.Interface().(type) {
+	case protoreflect.Message:
+		attrs := []rel.Attr{}
+		message.Range(func(desc protoreflect.FieldDescriptor, val protoreflect.Value) bool {
+			item, err := WalkThroughMessageToBuildValue(val)
+			if err != nil {
+				attrs = append(attrs, rel.NewAttr("@error_"+string(desc.Name()),
+					rel.NewString([]rune(err.Error()))))
```

Comment:
Is this necessary? If we fail to convert the value to a `rel.Value`, it seems as though the whole function should error out. What's the case where a conversion error would be acceptable?

This would simplify all arms of this switch as well.

Reply:
It can provide all errors in the `rel.Value` in one go.